### PR TITLE
packagekit: Show current package when reattaching to running update

### PR DESCRIPTION
**Pull request by martinpitt**
So far, watchTransaction() only updated the state on "Package" signals.
When reattaching to an already running transaction (on page reload, or
after relogin), this kept the "Initializing.." empty state until the
next Package signal arrived.

We can do better: The transaction object has `LastPackage` and `Status`
properties which we can read to get the data for the package which is
currently being updated. So factor out the code for reading the D-Bus
object properties and computing the state, and call it for both the
"Package" signal and the state of the transaction when attaching to it.

This will also make it easier for us to work on the "Applying updates"
page, as our tests create packages which hang forever -- and thus we
couldn't use page reloads to check the effect of code changes.

----

Also do some code modernization.

### CODE REVIEW

**Comment by martinpitt** [comment]
We can do better here: We don't need the two different code paths for the "initial" and "update after signal" cases for where to get the properties from. We can just always read them from the transaction. Updated.

**Comment by martinpitt** [comment]
There's some [bug on CentOS 8](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17517-20220701-095218-c911713a-centos-8-stream/log.html#300) which does not immediately set the correct properties after the `Package` signal. So I put back the more complicated "two code paths", as that seems to work fine. :man_shrugging:

File: pkg/packagekit/updates.jsx [hunk 1]
```diff
@@ -513,116 +513,106 @@ class RestartServices extends React.Component {
...
+    const [actions, setActions] = useState([]);
+
+    const updateForProperties = (status, packageId) => {
+        // small timeout to avoid excessive overlaps from the next PackageKit progress signal
+        PK.call(transaction, "org.freedesktop.DBus.Properties", "GetAll", [PK.transactionInterface], { timeout: 500 })
+                .then(([props]) => {
+                    const percent = props.Percentage.v;
+                    let remain = -1;
+                    if ("RemainingTime" in props)
+                        remain = props.RemainingTime.v;
```
Comment (cockpituous) [inline pkg/packagekit/updates.jsx:529]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17517-20220705-092548-cbb7d34f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/packagekit/updates.jsx [hunk 2]
```diff
@@ -513,116 +513,106 @@ class RestartServices extends React.Component {
...
-                                    atBottom = true;
-                            }
-
-                            this.setState({
-                                actions: newActions,
-                                percentage: percent <= 100 ? percent : 0,
-                                timeRemaining: remain > 0 ? remain : null
-                            });
+                    setPercentage(percent <= 100 ? percent : 0);
+                    setTimeRemaining(remain > 0 ? remain : null);
```
Comment (cockpituous) [inline pkg/packagekit/updates.jsx:546]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17517-20220705-092548-cbb7d34f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/packagekit/updates.jsx [hunk 3]
```diff
@@ -513,116 +513,106 @@ class RestartServices extends React.Component {
...
-                <div className="progress-main-view">
-                    <div className="progress-description">
-                        <Spinner isSVG size="sm" />
-                        <strong>{ PK_STATUS_STRINGS[lastAction.status] || PK_STATUS_STRINGS[PK.Enum.STATUS_UPDATE] }</strong>
-                        &nbsp;{lastAction.package}
-                    </div>
-                    <Progress title={timeRemaining} value={this.state.percentage} />
-                    {cancelButton}
+    const lastAction = actions[actions.length - 1];
+    const formatTimeRemaining = timeRemaining && timeformat.distanceToNow(new Date().valueOf() + timeRemaining * 1000);
```
Comment (cockpituous) [inline pkg/packagekit/updates.jsx:581]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17517-20220705-092548-cbb7d34f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/packagekit/updates.jsx [hunk 4]
```diff
@@ -1248,14 +1233,14 @@ class OsUpdates extends React.Component {
...
-                .fail(this.handleLoadError);
+                .catch(this.handleLoadError);
     }
 
     watchUpdates(transactionPath) {
         this.setState({ state: "applying", applyTransaction: transactionPath, allowCancel: false });
 
         PK.call(transactionPath, "DBus.Properties", "Get", [PK.transactionInterface, "AllowCancel"])
-                .done(reply => this.setState({ allowCancel: reply[0].v }));
+                .then(([prop]) => this.setState({ allowCancel: prop.v }));
```
Comment (cockpituous) [inline pkg/packagekit/updates.jsx:1243]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17517-20220705-092548-cbb7d34f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/packagekit/updates.jsx [hunk 5]
```diff
@@ -1294,7 +1279,7 @@ class OsUpdates extends React.Component {
                                        if ("AllowCancel" in notify)
                                            this.setState({ allowCancel: notify.AllowCancel });
                                    })
-                .fail(ex => {
+                .catch(ex => {
```
Comment (cockpituous) [inline pkg/packagekit/updates.jsx:1282]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17517-20220705-092548-cbb7d34f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/packagekit/updates.jsx [hunk 6]
```diff
@@ -1313,7 +1298,7 @@ class OsUpdates extends React.Component {
                     this.watchUpdates(transactionPath)
                             .then(() => {
                                 PK.call(transactionPath, PK.transactionInterface, "UpdatePackages", [0, ids])
-                                        .fail(ex => {
+                                        .catch(ex => {
```
Comment (cockpituous) [inline pkg/packagekit/updates.jsx:1301]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17517-20220705-092548-cbb7d34f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/packagekit/updates.jsx [hunk 7]
```diff
@@ -513,116 +513,106 @@ class RestartServices extends React.Component {
...
-    }
+                    // scroll update log to the bottom, if it already is (almost) at the bottom
+                    if (log && atBottom)
+                        log.scrollTop = log.scrollHeight;
+                });
+    };
 
-    componentWillUnmount() {
-        this._mounted = false;
+    useEffect(() => {
```
Comment (mvollmer) [inline pkg/packagekit/updates.jsx:554]: What about `useInit` from pkg/hooks.js?

**Review by mvollmer — APPROVED** [review]
I would prefer `useInit` to `useEffect`.  The added uncovered lines replace previously uncovered lines, so that's fine.

**Pull request merged**